### PR TITLE
Account for `use` statements when resolving paths in `kani::stub` attributes

### LINKED ISSUE #1866: Resolve names during stubbing
This issue affects how we treat the function/method names supplied in `kani::stub(<original>, <replacement>)` attributes. We currently treat `<original>` and `<replacement>` verbatim as strings, so that they have to exactly match a name we find in the HIR. This means that we are unable to find many functions/methods in foreign crates, that we do not support relative paths, and that we are unable to take into account `use ... as ...` statements.

We should instead correctly resolve names during stubbing, as well as support path constructs like `self` and `super`.

### LINKED ISSUE #1998: Take into account `use` statements when resolving stub paths
Currently, we do not account for `use` statements when trying to resolve paths appearing in `kani::stub` attributes. For example, this harness fails to compile:

```rust
fn foo() -> bool {
    false
}

fn bar() -> bool {
    true
}

mod my_mod {
    use super::*;

    #[kani::proof]
    #[kani::stub(foo, bar)]
    fn harness() {
        assert!(foo());
    }
}
```

We are not able to resolve either `foo` or `bar`. We should fix this, both to make easier to specify stubs, and to ensure that stubs aren't resolving to the wrong functions/methods.

### PR DESCRIPTION
**Pull request by aaronbembenek-aws**
### Description of changes: 

This PR follows up on #1999.

Currently, when we resolve paths appearing in `kani::stub` attributes, we do not take into account any `use` statements that might be in play. This PR fixes this gap by accounting for these types of imports:

- `use XXX;`
- `use XXX as YYY;`
- `use XXX::*`
- `extern crate XXX as YYY;`

### Resolved issues:

Resolves #1998 

Related #1866 
Related #1997

### Call-outs:

Some care had to be taken so that we handle glob imports correctly, as paths resolved through them essentially have a lower priority than paths resolved other ways.

### Testing:

* How is this change tested? New tests

* Is this a refactor change? No

### Checklist
- [X] Each commit message has a non-empty body, explaining why the change was made
- [X] Methods or procedures are documented
- [X] Regression or unit tests are included, or existing tests cover the modified code
- [X] My PR is restricted to a single feature or bugfix

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 and MIT licenses.

### CODE REVIEW

**Comment by aaronbembenek-aws** [comment]
@celinval FYI, this is ready for review, but a maintainer needs to start the CI job.

**Review by celinval — COMMENTED** [review]
Sorry, I haven't had time to go through the entire PR yet. FYI, I'll be off next week.

File: kani-compiler/src/kani_middle/resolve.rs
Comment (aaronbembenek-aws) [reply kani-compiler/src/kani_middle/resolve.rs:364]: We currently don't have the span available anywhere in this module, and at this point in the code we don't even have the full name of the original function we're trying to resolve. We could potentially pass this information around all the different resolution functions so that we can produce helpful error messages here. However, I think it would be cleaner to change the resolution functions to return a `Result` and then leave it up to the client of this module to produce error messages.

**Comment by aaronbembenek-aws** [comment]
Thanks for taking an initial look! I will also be off until mid-January.

File: kani-compiler/src/kani_middle/resolve.rs
Comment (celinval) [reply kani-compiler/src/kani_middle/resolve.rs:364]: You could wrap these functions into a type that holds that information so you can generate more helpful error messages. Not on this PR though. :)

**Comment by rahulku** [comment]
updates on this?

**Comment by celinval** [comment]
I don't think @aaronbembenek is back yet.

**Comment by aaronbembenek** [comment]
That’s right - I’m currently on vacation, but I’ll have some time to make
updates starting next week.

File: kani-compiler/src/kani_middle/resolve.rs
Comment (aaronbembenek) [reply kani-compiler/src/kani_middle/resolve.rs]: I changed this to the name of the root module.

File: kani-compiler/src/kani_middle/resolve.rs
Comment (aaronbembenek) [reply kani-compiler/src/kani_middle/resolve.rs:364]: Okay, sounds like we're holding off on making changes here until later. I think it probably makes most sense to go with the approach of returning a `Result`, so that this `resolve` module does not become specialized to stubbing, but can also be used in other parts of Kani.

**Comment by aaronbembenek** [comment]
Okay, I'm back now! Let me know what other changes to make...

**Comment by aaronbembenek** [comment]
Also, looks like I need someone to approve this workflow...

**Review by celinval — APPROVED** [review]
Thanks

File: kani-compiler/src/kani_middle/resolve.rs
Comment (celinval) [reply kani-compiler/src/kani_middle/resolve.rs:364]: BTW, my suggestion is to keep it independent from stubbing. I'm OK with returning a result too, and letting the user to decide whether the failure to resolve is an error or not.

My main point is that the message should give more context on the exact path that we are trying to resolve and in which level it has failed.

**Comment by aaronbembenek** [comment]
Thanks @celinval for taking a look! You might have to merge this for me (I don't think I have permissions to).

**Pull request merged**
